### PR TITLE
#539 Implement working functionality for the GET /program/:programAssessmentId/submissions/new route

### LINKED ISSUE #539: Implement functionality for the GET /program/:programAssessmentId/submissions/new route and accompanying service file functions
### Describe the Feature / Enhancement

The `GET /program/:programAssessmentId/submissions/new` route needs to be implemented and fully tested. The `GET /program/:programAssessmentId/submissions/new` route creates a new assessment submission for a program assessment if the participant is allowed to do so.

There needs to be a test coverage for the `GET /program/:programAssessmentId/submissions/new` route and the service file functions it needs. We need to test this route to ensure the code written for the test will behave as expected by the protocol we previously defined.

This issue continues the work from issue #515.

### Additional Details and Resources

This route requires the principal ID (the user ID) from the session information, this route expects a program assessment ID in the URL as a URL parameter, and expects nothing in the body of the request. The response should contain the following:

#### Facilitator

This should return an UnauthorizedError. Facilitators shouldn't have access to this route.

#### Participant

If the participant has not submitted the `max_num_submissions` for a curriculum assessment for this program assessment, and there exists no AssessmentSubmission currently in the "Opened" or "In Progress" states by this user for this program assessment, and the due date for this program assessment has not passed, return the following:

```typescript
res.json(
  itemEnvelope({
    curriculum_assessment: CurriculumAssessment (with 'questions' and 'answers' but not the correct answers),
    program_assessment: ProgramAssessment,
    principal_program_role: 'Participant',
    submission: a newly-created AssessmentSubmission
  })
);
```

If the participant has an AssessmentSubmission currently in the "Opened" or "In Progress" for this program assessment and the due date for this program assessment has not passed, return the following:

```typescript
res.json(
  itemEnvelope({
    curriculum_assessment: CurriculumAssessment (with 'questions' and 'answers' but not the correct answers),
    program_assessment: ProgramAssessment,
    principal_program_role: 'Participant',
    submission: the existing AssessmentSubmission
  })
);
```

Else, return a ForbiddenError.

```typescript
export class ForbiddenError extends HttpError {}
ForbiddenError.prototype.message = 'Access to this resource is not allowed.';
ForbiddenError.prototype.status = 403;
```

```typescript
else if (new Date() > due_date) {
  throw new ForbiddenError(`You may not create a new assessment submission because the due date has passed.`);
} else if (numSubmissions > max_num_submissions) {
  throw new ForbiddenError(`You have reached the maximum number of submissions for this assessment.`);
} else {
  throw new ForbiddenError();
}
```

### Correct Project Selected

- [X] This issue is linked to the appropriate project

### Labels

- [X] This issue has appropriate labels.

### PR DESCRIPTION
**Pull request by zeeeeli**
## Proposed changes

This pull request close #539 which includes 
* GET /program/:programAssessmentId/submissions/new route
* route test
* service file functions related to the route(`listParticipantProgramAssessmentSubmissions`, `createAssessmentSubmission`)
* service file function test

## Checklist

- [x] Related issue appears at beginning of pull request title with pound sign, and title describes the changes being proposed.
- [x] Related issue is linked in pull request description using [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).
- [x] The appropriate label has been chosen for this pull request.
- [x] The correct project has been selected for this pull request.
- [x] All commits in this branch, including merge commits, begin with the issue number and a pound sign.
- [x] Tests have been added, where appropriate; or, no tests are relevant for this pull request.
- [x] This pull request contains UI changes, and screenshots of those UI images appear below; or, this pull request contains no UI changes.

### CODE REVIEW

File: api/src/middleware/__tests__/assessmentsRouter.ts [hunk 1]
```diff
@@ -870,11 +877,128 @@ describe('assessmentsRouter', () => {
...
+            );
+            done(err);
+          }
+        );
+    });
+
+    it('should respond with a NotFoundError if the assessment id was not found in the database', done => {
+      const programAssessmentNotFoundId = 0;
+      mockPrincipalId(participantId);
+      mockGetProgramIdByProgramAssessmentId.mockResolvedValue([
```
Comment (seidior) [inline api/src/middleware/__tests__/assessmentsRouter.ts]: Would this not return an empty array in this circumstance?

**Review by seidior — CHANGES_REQUESTED** [review]
Fantastic work, @ZeyuanLi16! I only have a couple of notes:

- Four other tests I'd like you to check for this route:
  - `it("should return an error if no possible submissions remain for this participant and this assessment")`
  - `it("should return an error when attempting to create a new submission but one is currently opened or in progress")`
  - `it("should return an error when attempting to create a submission when the program assessment due date has passed")` and
  - `it("should return an error when attempting to create a submission for a program assessment that is not yet available")`
- I'm curious as to one thing in the expected structure of the logic of your route. You call `getProgramIdByProgramAssessmentId()` and then call `programAssessmentById()`; why not just call `programAssessmentById()` in the first place and extract what you need from its return result?

**Comment by zeeeeli** [comment]
EOD update
Add the route and some route tests according to the issue description.
The route had been fully manual tested.

Up next, finishing remaining route tests and add service file function tests.

File: api/src/middleware/assessmentsRouter.ts [hunk 2]
```diff
@@ -252,7 +255,146 @@ assessmentsRouter.get(
...
+      if (!programAssessment) {
+        next(
+          new NotFoundError(
+            `Could not find program assessment(ID ${programAssessmentIdParsed}).`
+          )
+        );
+        return;
+      }
+
+      if (new Date(programAssessment.available_after + 'Z') > new Date()) {
```
Comment (kaiyolau) [inline api/src/middleware/assessmentsRouter.ts:522]: can you explain a little bit about what it means? why do you create 'Z'?

Comment (seidior) [reply api/src/middleware/assessmentsRouter.ts:522]: JavaScript dates and UTC, I believe!

**Review by seidior — APPROVED** [review]
Great work on all the tests! It required minimal work to be able to get everything to pass. Well done, Ze!

**Pull request merged**